### PR TITLE
feat(ontology): refactor list of properties in resource class (DSP-1360)

### PR DESCRIPTION
**Pull request by kilchenmann**
resolves DSP-1360 (bug fix by returning resource class and properties in all languages)

incl. refactoring of cache and ontology editor in general. 

The ontology editor lists the resource classes with new list of corresponding properties; the design contains:

* gui order
* icon = property type (subclass and gui element); has a tooltip with more information
* label = property label with property comment in tooltip; 
  * in case of a link property it shows corresponding resource class
  * in case of a list property it shows corresponding list as a link
* cardinality = checkbox icons for "can have **multiple** values" and "value is **required**"

Screen Shot 2021-02-18 at 07 44 04

### CODE REVIEW

**Review by flavens — COMMENTED** [review]
- There are some index, e.g. 0) 1), that are doubled or tripled, or even missing, e.g. 0) -> 2)
- Is it normal behaviour to be able to click on the referred list but not on the linked resource?
- Maybe only on my local app: For the BEOL ontologies, I have an error message in the console.log: "Cannot read property 'label' of undefined" - `property-info.component.ts:95`

**Comment by kilchenmann** [comment]
> * There are some index, e.g. 0) 1), that are doubled or tripled, or even missing, e.g. 0) -> 2)

The Gui order is defined in the ontology; it's part of the res class cardinality. In case of missing or doubled numbers it's a mistake in the ontology. The test data are not well done.

> * Is it normal behaviour to be able to click on the referred list but not on the linked resource?

Yes. At the moment we do not have a native route for the linked resource class. I was thinking to highlight the linked class somehow. But this will take more time.

> * Maybe only on my local app: For the BEOL ontologies, I have an error message in the console.log: "Cannot read property 'label' of undefined" - `property-info.component.ts:95`

I will check this...

**Comment by mdelez** [comment]
> * Maybe only on my local app: For the BEOL ontologies, I have an error message in the console.log: "Cannot read property 'label' of undefined" - `property-info.component.ts:95`

I have the same console error. The BEOL ontology also has a lot of empty resource classes at the bottom for some reason. Maybe that's the issue @kilchenmann?

Screen Shot 2021-02-18 at 11 36 00

**Comment by kilchenmann** [comment]
Yes there's something weird with the ontology test-data in some projects. I will spend some time to do some investigations. Maybe someone of you can help me with that?

File: src/app/project/ontology/property-info/property-info.component.spec.ts [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+    beforeEach(() => {
+        testHostFixture = TestBed.createComponent(TestHostComponent);
+        testHostComponent = testHostFixture.componentInstance;
+        testHostFixture.detectChanges();
+    });
+
+    it('should create', () => {
+        expect(testHostComponent).toBeTruthy();
+    });
+});
```
Comment (mdelez) [inline src/app/project/ontology/property-info/property-info.component.spec.ts:90]: I would add a test here to ensure the component is receiving the correct data for `propDef`, `propCard`, and `projectcode` via the `@Input`s

File: src/app/project/ontology/property-info/property-info.component.ts [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+    templateUrl: './property-info.component.html',
+    styleUrls: ['./property-info.component.scss']
+})
+export class PropertyInfoComponent implements OnInit, AfterContentInit {
+
+    @Input() propDef: ResourcePropertyDefinitionWithAllLanguages;
+
+    @Input() propCard: IHasProperty;
+
+    @Input() projectcode: string;
```
Comment (mdelez) [inline src/app/project/ontology/property-info/property-info.component.ts:27]: projectcode -> projectCode

**Comment by kilchenmann** [comment]
I already found something: the empty resource classes doesn't have a label. They are all of type "Standoff"-something. Will try to fix it

**Review by mdelez — CHANGES_REQUESTED** [review]

File: src/app/project/ontology/property-info/property-info.component.ts (hunk 2 above)
Comment (kilchenmann) [reply src/app/project/ontology/property-info/property-info.component.ts:27]: it's the same as in other components. To be consistent, I have to change it everywhere. But I can do it...

**Comment by kilchenmann** [comment]
> I  have the same console error. The BEOL ontology also has a lot of empty resource classes at the bottom for some reason. 

I found the problem. The linked resource is defined in another ontology. The ontology editor is not yet prepared for this case. It will make the things more complex 🙈 ... but fine, I'll fix it.

**Comment by mdelez** [comment]
also, how come the package-lock.json file changed? Did you update some packages? I've also changed the file in my [PR #388](https://github.com/dasch-swiss/dsp-app/pull/388) because I needed to update the DSP-JS version and github does not merge these files for you because it is too large.

File: src/app/project/ontology/property-info/property-info.component.ts (hunk 2 above)
Comment (mdelez) [reply src/app/project/ontology/property-info/property-info.component.ts:27]: I've noticed it elsewhere as well. Maybe leave it for now and we can open a small PR to rename them everywhere all at once.

**Comment by kilchenmann** [comment]
It seems to be fixed now. I had to integrate new cache for `currentProjectOntologies`. Now we have the class information from other ontologies than the current one if in case a link property is connected with a class from another ontology.

**Comment by subotic** [comment]
great, so as soon as you merge this, please make a release and deploy it to test and staging. thanks.

File: src/app/project/ontology/property-info/property-info.component.spec.ts (hunk 1 above)
Comment (mdelez) [reply src/app/project/ontology/property-info/property-info.component.spec.ts:90]: @kilchenmann please add these assertions and then I will approve :)

Comment (kilchenmann) [reply src/app/project/ontology/property-info/property-info.component.spec.ts:90]: Will add it this evening.

Comment (kilchenmann) [reply src/app/project/ontology/property-info/property-info.component.spec.ts:240]: done in 8cf574e

**Review by mdelez — APPROVED** [review]

**Pull request merged**